src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSigner.cs

```diff
@@ -188,85 +239,128 @@ internal void CheckCertificateValue()
                 throw new PlatformNotSupportedException(SR.Cryptography_Cms_NoSignerCert);
             }
 
-            if (PrivateKey == null && !Certificate.HasPrivateKey)
+            if (_privateKey == null && !Certificate.HasPrivateKey)
             {
                 throw new CryptographicException(SR.Cryptography_Cms_Signing_RequiresPrivateKey);
             }
         }
 
-        internal SignerInfoAsn Sign(
-            ReadOnlyMemory<byte> data,
-            string? contentTypeOid,
-            bool silent,
-            out X509Certificate2Collection chainCerts)
+        private byte[] PrepareAttributesToSign(ReadOnlySpan<byte> contentHash, string? contentTypeOid, out AsnWriter newSignedAttrsWriter)
         {
-            HashAlgorithmName hashAlgorithmName = PkcsHelpers.GetDigestAlgorithm(DigestAlgorithm);
-            SignerInfoAsn newSignerInfo = default;
-            newSignerInfo.DigestAlgorithm.Algorithm = DigestAlgorithm.Value!;
-            byte[] dataHash;
+            List<AttributeAsn> signedAttrs = PkcsHelpers.BuildAttributes(SignedAttributes);
+
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+            writer.WriteOctetString(contentHash);
 
-            IncrementalHash hasher;
+            signedAttrs.Add(
+                new AttributeAsn
+                {
+                    AttrType = Oids.MessageDigest,
+                    AttrValues = new[] { new ReadOnlyMemory<byte>(writer.Encode()) },
+                });
 
-            try
+            if (contentTypeOid != null)
             {
-                hasher = IncrementalHash.CreateHash(hashAlgorithmName);
+                writer.Reset();
+                writer.WriteObjectIdentifierForCrypto(contentTypeOid);
+
+                signedAttrs.Add(
+                    new AttributeAsn
+                    {
+                        AttrType = Oids.ContentType,
+                        AttrValues = new[] { new ReadOnlyMemory<byte>(writer.Encode()) },
+                    });
             }
-            catch (PlatformNotSupportedException ex)
+            // else: TODO if we're in pure mode we *should* add a content type according to
+            // the SLH-DSA and ML-DSA spec. However, the only case when the content type is null
+            // is when we're countersigning, and RFC 5652 specifically states that
+            // countersignatures must not contain a content type. We'll leave it as is for now
+            // as countersignatures don't seem to be in the SLH-DSA CMS spec.
+
+            // Use the serializer/deserializer to DER-normalize the attribute order.
+            SignedAttributesSet signedAttrsSet = default;
+            byte[]? attributesToSign = null;
+            signedAttrsSet.SignedAttributes = PkcsHelpers.NormalizeAttributeSet(
+                signedAttrs.ToArray(),
+                encodedSignedAttrs => attributesToSign = encodedSignedAttrs);
```

Comment:
If this is the only call to NormalizeAttributeSet now, just change it to `out` the value.  Capturing via a lambda feels icky.

Reply:
Looking at current usage, everyone else passes null, which means making a proper overload, one that does the work and has an out, and the other that just sends the out to discard.